### ISSUE TITLE
Xbox Elite Controller not working yet.

### ISSUE THREAD

**Issue by chadthman** [360Controller/360Controller#160]
Just got the Elite Controller and unfortunately it appears to not be supported yet. Here is what it shows in system profiler if that helps:

``` 
Controller:

  Product ID:	0x02e3
  Vendor ID:	0x045e  (Microsoft Corporation)
  Version:	2.03
  Serial Number: <12 digit alpha-numeric number here>
  Speed:	Up to 12 Mb/sec
  Manufacturer:	Microsoft
  Location ID:	0x1a153000 / 11
  Current Available (mA):	1000
  Extra Operating Current (mA):	0

**Comment by TyanColte** [comment]
Same Problem here, also the standard controller with the latest firmware update is also not working. Please fix. I miss my controller.

**Comment by FranticRain** [comment]
Forgot to mention it since I added it before this issue was created (October 3rd) but this is resolved in #151.

**Comment by TyanColte** [comment]
@FranticRain do we have a release to go with this resolution?

**Comment by FranticRain** [comment]
I don't do releases, but since the pull request hasn't been accepted, there it is not in any current build. But as far as I know, we can't build because the build system is broken (see #163) and the driver is broken in the latest version of El Capitan (see #161 and #167.) As far as I can guess, the maintainers are waiting for Apple to stop screwing around with the driver API and stop breaking things.

**Comment by TyanColte** [comment]
@FranticRain well that's a bummer. When do you speculate a new release will be? I'm going to be going on a trip soon and I really would like all this to work. :+1:

**Comment by FranticRain** [comment]
Unfortunately, I can't speak for the rest of the dev crew. They get it done when things work and they have the time. If you're willing to manually install the driver, I can compile a bootleg version of the driver. I never have figured out how all of the build scripts fit together, but I do have a kext signing certificate. Let me know if you are interested.

**Comment by TyanColte** [comment]
@FranticRain sure, i'm up for testing it. let me know when you've got one for me. I'll go ahead and plop it in and let you know how it worked.

**Comment by FranticRain** [comment]
Okay, download the version @herrbischoff was nice enough to compile [here.](https://github.com/360Controller/360Controller/pull/151#issuecomment-158693976)

Then move the "360Controller.kext" into `/Library/Extensions`
You can get to this directory easily by using the `Go -> Go to Folder...` or `CMD + Shift + G` option in Finder.
Once it is there, click on it, and then make an alias in Finder with `File -> Make Alias` or `CMD + L`.
Move the Alias to `/System/Library/Extensions` and put it there.
Once that is placed, you should be good to go. You'll probably need to reboot before it'll work.

**Comment by TyanColte** [comment]
@FranticRain Awesome! I'll let you know if it works.

**Comment by FranticRain** [comment]
Ah, he didn't sign the kext. That's why its not working. I should have a kext signing certificate, but I recently reinstalled OS X and appear to have lost it. I'll do my best to recover it tomorrow.

**Comment by TyanColte** [comment]
@FranticRain I have a friend who may be able to accomplish this for me. Thanks for the help you've provided so far.

**Comment by TyanColte** [comment]
@FranticRain so i was finally able to get this installed, I had to allow kext-dev-mode=1 and disable System Integrity Protection on my entire mac. But it works, it loads the kext. It also recognizes the elite controller, but the controls are all wonky, moving a control stick around pushes random buttons and directions on the D-Pad, among other issues (maybe having to do with the ability to remap any control as a different control on your Xbox One or windows 10). It also still doesn't recognize the original Xbox one controller with the latest firmware update installed on it.

**Comment by FranticRain** [comment]
I swear if Microsoft changed the packet layouts on me again, so help me. I'll get Windows 10 set up and update my controller. Then I'll sniff the packets to see whats up.

**Comment by TyanColte** [comment]
Knowing M$ that wouldn't surprise me. They really don't want mac users using this stuff on their macs. Even though it's the best of all the controllers on the market.

**Comment by TyanColte** [comment]
@FranticRain Do you all have an IRC or anything I can get in to help contribute with. Chatting in comments like this is kind of an odd duck.

**Comment by FranticRain** [comment]
What is the product and vendor ID on your original Xbox One controller? That thing should be working on the latest version of the driver unless you have a third party controller.
As for chatting with comments, its just the typical fair for this kind of this. We don't have any secret dev chat or anything.

**Comment by TyanColte** [comment]
Lol, ok, i referenced the product and vendor IDs of my original xbone controller in #151 latest comment.

**Comment by TyanColte** [comment]
@FranticRain if you want, i can set up wireshark and usbpcap on my windows 10 machine and cap some packets for you. i'll send you the results of you send me some way i can do so.

**Comment by FranticRain** [comment]
Yes! If you could do that, it would be incredibly useful. Here's how:
Follow the "Following Along" section of [this website](http://www.definedbehavior.com/reverse-engineering-the-xbox-one-controller/). Once you are done there, skip down the to "The stuff magic is made of" section and sniff the packets. If you could just send me the whole capture files, that would be great. Additionally, if you could make one scan of you pressing buttons in a particular order on the controller, that would be very useful. The order the buttons used to be in was A, B, X, Y, LB, RB, View, Menu, Left Stick Click, Right Stick Click, Guide, Sync, Up, Down, Left, Right, Left Stick, Right Stick, Left Trigger, Right Trigger.

**Comment by FranticRain** [comment]
And if for some reason you are interested in looking at my notes for previous packet layouts, you can see those [here](https://www.dropbox.com/s/25dd5a6irn7jfqt/RAWPACKETS.html?dl=0).

**Comment by TyanColte** [comment]
@FranticRain Got the cap, can't upload it here. Will put it on my GDrive and send you a link.

**Comment by TyanColte** [comment]
@FranticRain https://drive.google.com/file/d/0B381VMfcesrldnFuOE1aSmpDREE/view?usp=sharing

**Comment by FranticRain** [comment]
@TyanColte According to that, the only packets being set from the computer to the controller is an init packet that is already in the driver. `05:20:10:01:05` and `05:20:0f:01:05` from your scan are similar to the sent driver's `05:20:03:01:00`. You'll notice the 3rd byte differs. That is just an increment byte that isn't important. The last byte is different though. I've never seen a trailing `05` before. Which controller was this?

**Comment by TyanColte** [comment]
That was the version 1 xbox one controller with the latest firmware update received from the xbox accessories app in windows 10. it's the controller without the 3.5 mm jack on the bottom. nothing else is connected to the controller, not a chat pad or anything.

**Comment by FranticRain** [comment]
Okay, can you modify your testing a little bit? Open up USBPcap with the controller plugged in. When it gives you the filters, unplug the controller. Select the filter that the controller was in, give it a name, and then plug the controller in. Give it a couple seconds and then quit scanning. Then let me see that one.

**Comment by TyanColte** [comment]
Do i need to do that using wireshark or just usbpcapcmd by itself?

**Comment by FranticRain** [comment]
Just USBPcapCMD. No need for Wireshark. You really only need Wireshark if you want to look at the data.

**Comment by TyanColte** [comment]
ok, will do. i'll get you a report asap

**Comment by TyanColte** [comment]
@FranticRain https://drive.google.com/file/d/0B381VMfcesrlVzZuaGZ3Q2tYbXc/view?usp=sharing is this what you needed?

**Comment by FranticRain** [comment]
Yeah. That's it, and its virtually identical to the existing sniffed packets. I don't see anything wrong here.

**Comment by TyanColte** [comment]
Don't forget i'm using that unsigned version of the driver that @herrbischoff compiled. all i know is that when i plug it in it vibrates to show that it's getting power and then nothing.

**Comment by TyanColte** [comment]
but even when i was using the beta 3 driver it still wasn't working

**Comment by FranticRain** [comment]
The controller vibrates when using the OS X driver? It should just light up the Xbox logo. Its not supposed to vibrate. It should only produce rumble on Windows.

**Comment by TyanColte** [comment]
It does vibrate, as soon as i plug it in. I don't know to what end it does this for. However the elite controller does the same thing but then it lights up.

**Comment by FranticRain** [comment]
The Elite controller vibrates and then lights up on OS X?

**Comment by TyanColte** [comment]
Yea, with that one driver that @herrbischoff compiled.

**Comment by TyanColte** [comment]
Could it have something to do with the fact that i was using the xone driver before this? i removed the kext after uninstalling the entry in system preferences. but i may have some left over files.

**Comment by FranticRain** [comment]
Xone-OSX only has two files. The Pref Pane and `Xone Controller.kext`

**Comment by TyanColte** [comment]
well those have been removed. so there's that. hmmmmmm. I'm going to try reinstalling the 360controller package and see if that helps. I'll uninstall the driver first. as long as i can get the original xbox one controller working, i will be ok until the elite controller becomes supported.

**Comment by FranticRain** [comment]
The 2013 Xbox One controller works with the driver as far as I know. I'm paranoid about updating my 2015 controller because I still want to play Fallout 4 on my Windows 8 machine. >_>

**Comment by TyanColte** [comment]
what about the 2013 xbox one controller with the latest firmware update? because that's what i have and it wasn't working.

**Comment by FranticRain** [comment]
I don't have any 2013 controllers with me right now. I'll try and update my 2015 one and see if it has problems too.

**Comment by TyanColte** [comment]
are you using your xbox one to update your controller?

**Comment by FranticRain** [comment]
No. My Xbox One is not with me at the moment. I plan to update through Windows 10.

**Comment by TyanColte** [comment]
Oh, well that's about the only other way you can update it. I just reinstalled the 0.15-B3 package and rebooted my mac and still the controller isn't recognized.

**Comment by FranticRain** [comment]
Just updated my 2015 controller. It still works perfectly with OS X. There is a slight vibration now, but it runs just fine.

**Comment by TyanColte** [comment]
So what's the deal with mine? what am i doing wrong?

**Comment by FranticRain** [comment]
I honestly have no idea. Everything looks good to me.

**Comment by TyanColte** [comment]
would it matter what cable i'm using?

**Comment by FranticRain** [comment]
Yes, but only if you are using a data cable vs a charge cable. If you are getting vibration and it shows up in the system profiler, you are using a data cable.

**Comment by TyanColte** [comment]
Argh, when do you think you'll be able to get your hands on a 2013 controller, update it and see if it has the same problems as mine?

**Comment by FranticRain** [comment]
Not for a while. It would be after Thanksgiving.

**Comment by TyanColte** [comment]
Darn, well that's no good. I've gotta think of some other way. If only i could downgrade the firmware on my 2013 controller. I don't use the chat pad with it or audio or anything. I just updated it because i like to have up-to-date stuff and since i don't own an Xbone, when they released the xbox accessories app on windows 10 i was like "w00t" and then i was like "crap..."

**Comment by FranticRain** [comment]
Okay, lets just test this from scratch again, just to make sure.
Get your 2013 controller. Remove all accessories from the controller. Remove all batteries from the controller. Before plugging the controller in, run `kextstat | grep -i 360Controller`. This command shouldn't output anything.

Open the console. You can access this from `Applications -> Utilities -> Console`. Then in the left hand panel, click on system.log. Make note of what the last message you see is. We will be interested in the output after we connect the controller.

Now plug your controller in, observer and note its behavior. Now run `kextstat | grep -i 360Controller` again. This time it should have some sort of output. Something like this:
```
141    0 0xffffff7f832f9000 0x9000     0x9000     com.mice.driver.Xbox360Controller (1.0.0d15) 9DF80CD4-09D9-3E65-8CA0-AE3D42E0D206 <39 31 5 4 3>
```
If you don't see that, tell me what the last items listed are when you run the `kextstat` command by itself. I don't need the full output, just the last couple. Additionally, if the kextstat grep doesn't return anything, check and make sure you can see the controller in the system profiler `About This Mac -> System Profiler -> USB -> Controller`

Check the output of the console for any message prefixed by `start - `. If you see anything like that, post the recent contents of your system.log.

**Comment by TyanColte** [comment]
@FranticRain I'll do just that as soon as I get home from work. About 6:00 American Central Time. So about 6.5 hours from this posting.

**Comment by TyanColte** [comment]
@FranticRain Ok, so i did everything you said when i plugged in my controller i got two weird things in my console system.log

    Nov 23 18:21:49 Christophers-MacBook-Pro kernel[0]: start - unable to open device
    Nov 23 18:21:49 Christophers-MacBook-Pro cfprefsd[136]: BUG in libdispatch: 15B42 - 1631 - 0x0

and when i ran the `kextstat | -i grep 360controller` after plugging in the controller i received this output:

    148    0 0xffffff7f82da7000 0x9000     0x9000     com.mice.driver.Xbox360Controller (1.0.0d15) 9DF80CD4-09D9-3E65-8CA0-AE3D42E0D206 <43 32 5 4 3>

let me know what you think.

**Comment by TyanColte** [comment]
@FranticRain i don't think the `BUG in lib dispatch: 15B42 - 1631 - 0x0` has anything to do with it because i see similar things farther up in the log but it was the two instant things that came up when I plugged in the controller.

**Comment by TyanColte** [comment]
OK, something different happened that time in the log when i plugged in my controller

    Nov 23 18:40:07 Christophers-MacBook-Pro kernel[0]: start - unable to open device
    Nov 23 18:40:08 Christophers-MacBook-Pro kernel[0]: kIOMessageServiceIsRequestingClose

that second message came up

**Comment by TyanColte** [comment]
And this came up in the log when i unplugged it.

    Nov 23 18:40:32 Christophers-MacBook-Pro kernel[0]: kIOMessageServiceIsTerminated
    Nov 23 18:40:32 Christophers-MacBook-Pro kernel[0]: Stopping
    Nov 23 18:40:32 Christophers-MacBook-Pro kernel[0]: Free!

**Comment by FranticRain** [comment]
So what this means is that the driver matched to your controller, and started loading. But it failed to open the IOUSBDevice, which is a weird thing to have happen. The "start - " stuff is really the only part that matters in these logs. The rest of this data is typical. I'll keep looking into this, but I don't for the life of me know why the open would fail.

**Comment by TyanColte** [comment]
Thanks for looking into this @FranticRain I'm glad you're on top of it. I'd be lost without this github community. :+1:

**Comment by FranticRain** [comment]
Could you check out the "All Messages" portion of the console and see if there is anything prefixed with `handleOpen - `? I have no idea where USBLog prints out to, but there should be some printout there.

**Comment by TyanColte** [comment]
No dice.

**Comment by FranticRain** [comment]
I really, really don't know.

@MaddTheSane @RodrigoCard 
Any idea why a 2013 Xbox One controller would fail on line 392 of `_Controller.cpp` at `if(!device->open(this)) {`? The call returns false and then calmly fails, no panic.

**Comment by TyanColte** [comment]
Would reinstalling OS X help? I've been thinking about doing it anyway because i've acquired some cruft from trying to install crap on it.

**Comment by FranticRain** [comment]
Most likely not. But you never know, I guess.

**Comment by TyanColte** [comment]
So i guess there's no answer to this yet. I'm going to give it a couple more days and then i'm going to wipe my macbook and reload it and see if that fixes the issue. It may not but at least i'll have a fresh EL Capitan to work with.

**Comment by TyanColte** [comment]
@FranticRain I understand we've been working hard on this issue. But I think i found a solution, and it doesn't require an XBox controller whatsoever. I dug out my old ON-Live universal controller and hooked it up to my mac via bluetooth. Thanks to steam i was able to configure the controller to work just like an XBox controller. Now all the games that use xbox controllers work just fine. I'll stick to my microsoft hardware on microsoft devices from now on. this controller works wirelessly and i can pack it around with me. Thanks for all your help. If you do happen to come up with a solution go ahead and let me know. I'll be checking back periodically just for fun.

**Comment by FranticRain** [comment]
Thanks for the update. Ironically, I just got done updating a 2013 Xbox One controller and it works perfectly. I honestly have no idea what your problem is, and I can't reproduce it whatsoever. But at least you have a working controller now. Thats what is important.

**Comment by TyanColte** [comment]
@FranticRain well that tears it, i'm going to be reinstalling OS X on my laptop tonight. I found that some games don't quite work with the online controller yet. So this gives me another reason to do this. Hopefully this fixes the issues. Thanks for all the help. I'll let you know how it goes.

**Comment by TyanColte** [comment]
@FranticRain IT WORKED!!!!!! I was going to be very mad at myself if reloading OS X didn't fix the issue but it DID. I'm SO HAPPY!!! now if i could just get my elite controller working then all will be right with the world. But i'll wait for an official release for that. Thank you for all your help You've been awesome and i hope i can refer to you in the future if i have any issues or need something added to the repo. Thanks again.

**Comment by FranticRain** [comment]
@TyanColte Awesome! I'm glad to hear it worked out. I just hope the button mapping issues you experienced with the Elite controller were merely a symptom of your problem and not a full blown alternate mapping. If it does turn out to be an issue, I'll have to call on you to sniff the packets since you clearly know how to do it now. Thanks for sticking it out with me, and I hope we can get the Elite controller working for you soon.

**Comment by chadthman** [comment]
It's not just him I get the same issues with the elite controller.

**Comment by TyanColte** [comment]
@chadthman using the driver @herrbischoff compiled?

**Comment by chadthman** [comment]
I actually just modified my info.plist to recognize the controller.

**Comment by TyanColte** [comment]
@chadthman OH, well there's that lol

**Comment by chadthman** [comment]
Basically made the change that is in commit dfaa22b

**Comment by TyanColte** [comment]
@chadthman and you're does the eratic crazy button presses also?

**Comment by chadthman** [comment]
Yep and if you press the center button that lights up it appears to reset itself.

**Comment by TyanColte** [comment]
@chadthman exactly what happens to me.

**Comment by TyanColte** [comment]
@chadthman when i edit the info.plist the kext goes unsigned and unloads itself. It won't load the kext and messes up my driver. So i'm unable to test that by just modifying the info.plist.

**Comment by chadthman** [comment]
@TyanColte You might need make sure that you're machine is using kext-dev-mode=1 and probably disable SIP. But first you can see if refreshing the kernel cache will work for you. I would download [Kext Utility](http://cvad-mac.narod.ru/index/0-4) and run it then restart. FYI it will also add kext-dev-mode=1, -v, and rootless=0 boot flags by default so before you run it you can change them if you edit the KU_config.plist inside the app.`

**Comment by TyanColte** [comment]
@chadthman yea i'm pretty sure i'm not going to disable all those security measures again this time. I'll just wait until they release a signed version and merge the pull request with the main branch.

**Comment by TyanColte** [comment]
@FranticRain If you want me to i can get you a usb cap of the elite controller to see if these frantic button presses are as a result of packet changes between the regular xbone controller and the elite

**Comment by FranticRain** [comment]
@TyanColte Yeah. If you could do that it'd be great. Make sure to grab the start packet, so start scanning before plugging the controller in. Then make sure to press the buttons one at a time so I can figure stuff out. You might even try rebinding the controller in the app while sniffing. Also make sure to press the paddles while sniffing.

**Comment by TyanColte** [comment]
@FranticRain i can do that, does it matter if i use the usb cable or the wireless adapter? I'd assume the usb cable is going to be what you want. But I figured i'd ask either way.

**Comment by FranticRain** [comment]
Oh boy! If you could sniff the Elite controller wired and the 2013 controller wirelessly, that would be awesome! We need the init information for the wireless adapter as well, so make sure to sniff before connecting.

**Comment by TyanColte** [comment]
https://drive.google.com/file/d/0B381VMfcesrlaG5fTjl3WjZuc2c/view?usp=sharing Here's my Elite controller wired sniff button press order is as follows:

A, B, X, Y, LB, RB, View, Menu, Left Stick Click, Right Stick Click, Upper Left Paddle, Upper Right Paddle, Lower Left Paddle, Lower Right Paddle, Preset Switch toggle to slot 2, then back to slot 1, Left Stick X Axis, Left Stick Y Axis, Right Stick X Axis, Right Stick Y Axis, Left Trigger, Right Trigger. Opened Xbox Accessories App, Remapped Upper left Trigger to LBumper, Remapped Upper Right Trigger to RBumper, Saved configuration to Slot 1 on the controller. let it sit and collect data for a few seconds. i believe that's the exact order. If something doesn't seem right, i can run it again. I'll do the wireless one next starting with the 2013 controller.

**Comment by TyanColte** [comment]
https://drive.google.com/file/d/0B381VMfcesrlNWZwRkpoTXJBaDA/view?usp=sharing

Wireless adapter capture with the 2013 controller. order of operations include: 

Plugging in adapter to usb port, Pressing sync button on adapter, pressing sync button on controller, completing sync, pressing buttons in order of A, B, X, Y, LB, RB, LS, RS, Guide, DPadUp, DPadDown, DPadLeft, DPadRight, LS X Axis, LS Y Axis, RS X Axis, RS Y Axis, Left Trigger, Right Trigger, Holding down guide button to turn off controller, Unplugging Adapter. Hope this helps.

**Comment by TyanColte** [comment]
@FranticRain Both sniffs are posted above, Btw, is there a way i can give you my email in case you need anything else, (Such as chatpad sniff or audio device sniff) where i won't be announcing it to the entire github community?

**Comment by FranticRain** [comment]
@TyanColte I think letting everyone see the sniffs is good for the open source community, but if you want to keep them private, you can email me at my github username at gmail.

Its going to take me a while to do anything actionable with this data. I'm out of town for Thanksgiving, so things are a bit slow.

**Comment by TyanColte** [comment]
@FranticRain I didn't mean keeping the sniffs private, i just meant a direct contact method for you in case you need anything else from me. I'll post it on github but this way i don't have to keep checking github and i'll get a notification when an email comes in. That's all, i want the community to see the sniffs, it'll only be a good thing for the project. just meant a way you can direct contact me if you need anything and i'm not checking github as regularly as i was when the 2013 controller screwed up on me.

**Comment by FranticRain** [comment]
@TyanColte Sounds good. Just send me an email and I'll get back with you there if you need anything.

**Comment by TyanColte** [comment]
@FranticRain Just sent you an email. You'll recognize it. Thanks a lot for all your help.

**Comment by TyanColte** [comment]
https://drive.google.com/file/d/0B381VMfcesrlRmlKN0tPWGVfdFU/view?usp=sharing Second version of the Elite Controller pcap. I took the sticks off and hopefully that fixed the isues with the stick spam. Let me know how it looks @FranticRain

**Comment by FranticRain** [comment]
@TyanColte Your left stick still has a heck of a drift, but the right one is rock solid. I just wrote a script to clean it up instead. From the best of my understanding, it looks like you pressed: A, B, X, Y, LB, RB, View, Menu, Left Stick Click, Right Stick Click, D-Pad Up, D-Pad Down, D-Pad Left, D-Pad Right, Upper Left Paddle (A), Upper Right Paddle (B), Lower Left Paddle (X), Lower Right Paddle (Y), [Something Happens Here that I can't see], Upper Left Paddle (LB), Upper Right Paddle (RB), Lower Left Paddle (X), Lower Right Paddle (Y). I'm ignoring any stick and trigger movement.
So for the part I couldn't see, did you change the switch on the controller from preset 2 to preset 1?

From here on out, I'm going to use [] notation to refer to the individual bytes as parts of an array. Therefore the very first byte in a set would be [0]. So for controller data that would be a 0x20.

So here is what I learned. The Elite controller sends a double length packet. The first part, [3] -> [17] packages the button data exactly the same as a normal controller. This data is the modified data. So things like the paddle presses look like normal button presses here.
But from [18] -> [31], the true data is shown. For example if the upper left paddle was mapped to A, [4] would be 0x10, while [18] would be 0x00. [4] reports that A is pressed, but [18] tells you that the physical A button was not actually pressed.
Byte [32] is a very interesting thing though. When you are just pressing random buttons on the controller, its 0x10. But when you use the paddles, it shifts. 0x11 for ULP, 0x12 for URP, 0x14 for LLP, and 0x18 for LRP. Then, you changed something and they lost their second tens place. 0x01 for ULP, 0x02 for URP, 0x04 for LLP, and 0x08 for LRP. Then when a paddle isn't being pressed, it is 0x00 instead of 0x10.

Additionally, it looks like rebinding takes place completely on the controller, which is fine by me.

As for why the driver doesn't work with the controller right now, that's pretty clear. Because this packet is suddenly double the size, we aren't prepared to handle it. I have no idea how you managed to get any sort of clicks out of it. Right, wrong, or otherwise.

Currently, on Line 537 of Controller.cpp, we check packet type and length with:
```
if ((report->header.command==0x20) && report->header.size==(sizeof(XBOXONE_IN_REPORT)-4)) {
```
`sizeof(XBOXONE_IN_REPORT)` is 18, subtract 4 gives us 14 or 0x0E, which is what a normal packet gives us. But an Elite controller gives us 0x1D or 29, so this line should always fail. When I get back to a controller, I'll mess around with doubling the length and see what I get.

**TL;DR** Elite controller be crazy. Will fix it.

**Comment by TyanColte** [comment]
@FranticRain sounds like you've got it figured out. And that tells me something else interesting i was wonder, You're correct that i did switch the preset from preset 2 to preset 1 when i did the part you couldn't see. But that tells me that all the settings you make in the xbox app are saved completely in the controller which is freaking awesome, and that you can't map the paddles inside the mac driver which completely sucks. To do that you'd have to write an entirely different ui for the driver specially for the elite controller and know exactly how it writes the data and saves it to the controller. I suppose that's fine. However it does make me happy that any changes i make on the controller from the xbox app, including saving presets to the toggle 1-2 configs will translate to the mac when the driver does get fixed. As for the stick drift on the left stick, I'm not really sure why that's happening. I'm going to try setting a deadzone in the xbox app and seeing if that helps, i don't notice it in any games tho. As for the larger packet size that may make sense because my theory is the controller sends 2 values, the button you actually pressed and the value of the button that was changed. That's just a theory so i'm not sure if it's correct. Hopefully all this info makes for a much better experience on the mac from the xbox elite controller. Thanks again.

**Comment by natbro** [comment]
not entirely clear from this trail to me - is the existing wireless adapter init path working for you @TyanColte & @FranticRain using the new XBox One Wireless Adapter (pid 0x02e6, vid 0x045e) against the Elite? or are you pairing an Elite with one of the old wireless adapters (pid 0x0719, vid 0x045e and pid 0x0291, vid 0x045e)? my win10 caps did not look like yours? also, if that was what you were doing, had you already paired the Xbox One Wireless Adapter and the Elite on a Windows machine before putting them into the Mac?

**Comment by FranticRain** [comment]
@natbro I don't know all of the details, but it is supposed to be a 2013 Xbox One controller connecting to a new Xbox One Wireless Adapter on Windows 10. I haven't looked at it at all, so I can't speak to any of the data contained in the sniff.

**Comment by chadthman** [comment]
Yea the Xbox One Wireless Adapter does not working in osx at all yet.

**Comment by TyanColte** [comment]
@FranticRain Any luck getting your cert back so you can build a signed version of this driver so i can test it for ya?

**Comment by FranticRain** [comment]
@TyanColte Not yet. They said one business day, and its been about one business day.

**Comment by TyanColte** [comment]
@FranticRain Ah, well good luck to ya. I'm looking forward to testing it. Have you made any progress on the elite controller's double packet size issues?

**Comment by FranticRain** [comment]
@TyanColte I've implemented a fix, but have been unable to test it, due to aforementioned signing issues.

**Comment by TyanColte** [comment]
Of course, Unfortunately I don't feel like compromising the security of my mac to test it. I guess we'll just wait until you hopefully get your cert back. Does it look to be a definite thing or a maybe situation?

**Comment by FranticRain** [comment]
Honestly, I haven't gotten anything back yet. So I have no clue. Anything could happen.

**Comment by TyanColte** [comment]
@FranticRain i think i figured out originally why my original xbox one controller wasn't working with my mac before i wiped it. I've been reading through the issues and whatnot just for fun and i came across somebody who had a similar problem as me. I realized very quickly that the reason it wasn't working for me was because i had your xone driver installed and i didn't uninstall it properly (I deleted the kext and removed the settings entry) before installing 360controller, i didn't know how there was a cmd that did all that. It's not a big deal but i thought you may like to know what i think happened knowing what i know now.

**Comment by FranticRain** [comment]
@TyanColte Interesting. I thought shipping something called "uninstall.command" next to the installer would have been obvious enough to people, but I guess it wasn't. I guess the vast majority of users have no idea how .kexts on OS X work, so it's not surprising that people wouldn't know to look for a deeper uninstall process than just removing the preference pane.

It's also a little funny, you were digging around in `/Library/Extensions` to install the `360Controller.kext` but you didn't notice `Xone Driver.kext`. That's pretty funny. If you had just noticed it and deleted it, things would be very different.

**Comment by FranticRain** [comment]
@TyanColte I got an update. I have to submit for a new certificate altogether. It could take months at best. Sorry. I'm turned off code signing and I am testing my changes locally, hopefully someone else can sign this thing for you once I get the bugs worked out.

**Comment by FranticRain** [comment]
@electric-monk Is it possible to check product and vendor IDs within the `start` function? I'm thinking about creating a subclass of `XboxOneControllerClass` for the Elite controller, but so far we tell the differences between controllers based on the interface data. I think we could set the `bAlternateSetting` in the `info.plist` but I have no clue how that works. Either way, I need to find the difference between an Xbox One controller and an Xbox One Elite controller, but I'm not quite sure how. Any ideas?

**Comment by electric-monk** [comment]
@FranticRain You should be able to read out the vendor and product IDs in the start function, yes - what you'd do is in the Info.plist, just as we have all the millions of brands of 360 controller. We just don't need that info as they're all the same for the 360 ;) When start is called you know your parent node, which should be some USB driver, so you can just query it's vendor/product ID properties.

Regarding the "alternate setting" thing, I think it refers to the optional USB thing where a device can have multiple descriptors that a driver/OS can use. If they're changing it with regards to these specific controllers, then I assume it will just be a matter of plopping it in alongside vendor/product IDs and it'll work, but I'd need to look into it in more detail to be sure.

**Comment by TyanColte** [comment]
@FranticRain well that's no good. I'll just disable rootless on my mac to allow the unsigned kext to work. At this point I just want to help out and test it. give me a buzz on email when you get a test release ready for me to play around with. Hopefully @RodrigoCard can get an official release out soon with those changes after we test them as good. Thanks a bunch.

**Comment by natbro** [comment]
@FranticRain easy to get the VID & PID to special-case certain things or set up an ivar flag to distinguish the elite for later. the 360ControllerClass overrides several superclass methods like `newVendorID` and `newProductID` so you can just call these directly to get those values. technically aside, i'm 99% sure these and several others don't need to be overridden - they are just repeating base-class behavior - so if you add `#define super IOHIDDevice` just below `OSDefineMetaClassAndStructors(...)` you can just call `super::newVendorID()` and `super::newProductID()`. you do need to perform this *after* `IOHIDDevice::start(provider)` in `start` (which by convention should be `super::start(provider)`, this is the point of the `#define super IOHIDDevice` kernel convention)

**Comment by natbro** [comment]
obviously going to need to alter `newReportDescriptor` when you have an Elite and I would suggest also altering `newProductString` - a frighteningly common recognition trick by games on PC and by extension probably to naive Mac ports is to look for "XBox 360" in the product string as a way to detect a complex gamepad rather than actually grok the input report description in some way - I suspect folks will eventually do this for Elite as well, so either matching the Elite's string when you know it or passing that call through to `super` to let it get picked off the device descriptor would be good.

**Comment by FranticRain** [comment]
The Elite controller should be supported in the most recent release. Let me know if you have any issues.
